### ISSUE TITLE
Easiest way to update coordinates?

### ISSUE THREAD

**Issue by hainm** [molstar/molstar#1375]
Hi

I am resuming my work to support molstar in jupyter lab/notebook.
What's the equivalent code to NGL's one?
I've been spending 1/2 day to search for the way but the molstar's API is too complicated for me to figure out. 
(Here is the code of how to use molstar in jupyter: https://github.com/nglviewer/nglview/blob/c3c08f617493e72c8d17b150c7758876066ac704/js/src/molstarview/widget.ts#L70-L83)

```
    updateCoordinates(coordinates, model) {
        // coordinates must be ArrayBuffer (use this.decode_base64)
        var component = this.stage.compList[model];
        if (coordinates && component) {
            var coords = new Float32Array(coordinates);
            component.structure.updatePosition(coords);
            component.updateRepresentations({
                "position": true
            });
        }
    }
```

Thanks

PS: I guess my question is there a way to avoid writing hundred lines of code?

**Comment by dsehnal** [comment]
You can define a `WithConformation :: Model -> Model` transform that takes the custom coordinates and applies them to the parent model which you "insert" as a child of the loaded model. Rest of the state reconciliation will happen automatically. It would work similarly to the existing `CustomModelProperties` transform.

Should be 20-40 lines of code :) I am traveling now but I can help over the weekend/early next week.

**Comment by hainm** [comment]
> Should be 20-40 lines of code :)

I can't still write them. Sorry. Feel like swimming in the ocean now. :D

**Comment by dsehnal** [comment]
Yeah, I can definitely understand that :) I wish I was able to write code that has NGLs simplicity combined with Mol*'s flexibility and performance scaling while documenting it in a way that is easy to understand quickly... but I don't know how to do that.

But if you are still interested in moving ahead with the Jupyter wrapper, I can help later as I wrote.

**Comment by hainm** [comment]
Yes I am very interested in molstar in jupyter. I will wait for your code.
Thanks.

**Comment by hainm** [comment]
For the record, I already implemented the initial work in nglview for
molstar
https://github.com/nglviewer/nglview/pull/1146

**Comment by hainm** [comment]
By the way, +1 for this post below:
<img width="950" alt="image" src="https://github.com/user-attachments/assets/4cee198c-9489-47a5-9d3c-57bfca057f71">

https://github.com/molstar/molstar/issues/557#issuecomment-1245231197

**Comment by dsehnal** [comment]
Yes, have been thinking about higher level API for some time. Just came back from PDBe visit and we were also talking about building new UI as well.

Regarding the Jupyter integration, have you considered supporting passing MolViewSpec states too? https://molstar.org/mol-view-spec/

**Comment by hainm** [comment]
> > Regarding the Jupyter integration, have you considered supporting passing MolViewSpec states too? https://molstar.org/mol-view-spec/
> 
> @dsehnal yes I do (with other things too). So below is my plan

In term if I was not aware of that specs package before you mentioned it. After you did, I have the plan.

**Comment by dsehnal** [comment]
https://github.com/molstar/molstar/pull/1378 Hopefully this will help, added docs with example usage too.

**Comment by hainm** [comment]
@dsehnal so basically I will need below code? 
- Also, if I already create the `hierarchy` somewhere, how can I get it from `this.plugin`? 
- And if I add two PDB files as two structure (or trajectory, not sure a good moltar term here). And I want to update coordinates for 2nd structure, how I could get 2nd `hierarchy`? 

```javascript
const updateCoords = async (plugin, coords) => {
    const hierarchy = await this.plugin.builders.structure.hierarchy.applyPreset(trajectory, 'default');
    const { x, y, z } = coords // [3, n_atoms]
    const coordinatesNode = await plugin.build().to(hierarchy!.model).insert(ModelWithCoordinates).commit();

    await plugin.build().to(coordinatesNode).update({
        atomicCoordinateFrame: {
            elementCount: x.length,
            time: { value: 0, unit: 'step' },
            xyzOrdering: { isIdentity: true },
            x: x,
            y: y,
            z: z
        }
    }).commit();
}
```

**Comment by dsehnal** [comment]
- You can only insert the "coordinates" node once, so repeated calls to the update trajectory will need to have extra logic (you can use `plugin.state.data.selectQ` for example to check if the coordinates node exists, say with combination of subtree/withTransformer functions). I was considering adding a helper function for "insert once", but decided to skip that for now as that use case isn't very common and can be handled in "user code"
- can use `plugin.managers.structure.hierarchy.current` to access all loaded structures. Can also keep the result of the `applyPreset` result in your plugin instance in a class field somewhere, including adding the "coordinates" node just once when the structure is loaded. Perhaps you can also have a 2nd wrapper class for each loaded structure in your plugin created when it's added that also maps to a python API and has methods that ensure the coordinates node is unique etc.

**Comment by hainm** [comment]
Thanks. I will try once you publish the new change.

**Comment by hainm** [comment]
Oh why does the structure don’t have coordinates node as default? The
existing coordinates after loading the pdb file is already there. Why can’t
we use that to update coordinates?

**Comment by hainm** [comment]
> Oh why does the structure don’t have coordinates node as default? The existing coordinates after loading the pdb file is already there. Why can’t we use that to update coordinates?
> […](#)
> On Sun, Dec 8, 2024 at 10:22 AM Hai Nguyen ***@***.***> wrote: Thanks. I will try once you publish the new change.

something like below? :D 
```
structure.updateCoordinates(coords_array)
```

**Comment by dsehnal** [comment]
> Oh why does the structure don’t have coordinates node as default?

Because the Mol* state management works in a similar way React does. You define declaratively way what needs to happen and the engine behind that performs the state reconciliation. This approach makes things like taking state snapshots and interpolation between states possible and quite generic. The tradeoff is a more complicated API.

**Comment by hainm** [comment]
Got it. Thanks.

**Comment by hainm** [comment]
@dsehnal Would you mind looking at the "FIXME" in below code? I am able to see the coordinates updated but I am pretty sure the code is not right. Thank you.

```python
async updateCoordinates(coordinates: any, modelIndex: any) {
        const n_atoms = coordinates.length / 3;
        const x = new Float32Array(n_atoms);
        const y = new Float32Array(n_atoms);
        const z = new Float32Array(n_atoms);

        for (let i = 0; i < n_atoms; i++) {
            x[i] = coordinates[i * 3];
            y[i] = coordinates[i * 3 + 1];
            z[i] = coordinates[i * 3 + 2];
        }
        const plugin = this.plugin;

        // Check if the coordinates node already exists
        const existingNode = plugin.state.data.selectQ(q => q.byRef('coordinatesNode'))[0]; // FIXME
        console.log("this.plugin.managers.structure.hierarchy.current.trajectories", this.plugin.managers.structure.hierarchy.current.trajectories);
        if (!existingNode) {
            console.log("Insert the coordinates node if it doesn't exist");
            var hierarchy = this.plugin.managers.structure.hierarchy.current.trajectories[modelIndex]; // FIXME: is this correct?
            var model = hierarchy.models[modelIndex]; // FIXME: is this correct?
            const coordinatesNode = await plugin.build().to(model.cell).insert(ModelWithCoordinates).commit();

            await plugin.build().to(coordinatesNode).update({
                atomicCoordinateFrame: {
                    elementCount: x.length,
                    time: { value: 0, unit: 'step' },
                    xyzOrdering: { isIdentity: true },
                    x: x,
                    y: y,
                    z: z
                }
            }).commit();
        } else {
            console.log("Update the existing coordinates node");
            await plugin.build().to(existingNode).update({
                atomicCoordinateFrame: {
                    elementCount: x.length,
                    time: { value: 0, unit: 'step' },
                    xyzOrdering: { isIdentity: true },
                    x: x,
                    y: y,
                    z: z
                }
            }).commit();
        }
    }
```

**Comment by dsehnal** [comment]
How about something like this:

```ts
class PluginWrapper {
    plugin: PluginContext;

    // ...

    async load(data): ModelWrapper {
        const model = new ModelWrapper(this);
        await model.load(data);
        return model;
    }
}

class ModelWrapper {
    get plugin() {
        return this.parent.plugin;
    }

    model: StateObjectSelector<PluginStateObject.Molecule.Model> | undefined;

    async load(data) {
        // ...
        const preset = await this.plugin.builders.structure.hierarchy.applyPreset(trajectory, 'default');
        this.model = preset?.model;
    }

    async updateTrajectory(coordinates) {
        if (!this.model) return;

        const node = this.plugin.state.data.selectQ(q => q.byRef(this.model.ref).children().withTransformer(ModelWithCoordinates))[0];
        const { x, y, z } = coordinates;
        if (!node) {
            await this.plugin.build().to(this.model).insert(ModelWithCoordinates, {
                atomicCoordinateFrame: {
                    elementCount: x.length,
                    time: { value: 0, unit: 'step' },
                    xyzOrdering: { isIdentity: true },
                    x: x,
                    y: y,
                    z: z
                }
            }).commit();
        } else {
            await this.plugin.build().to(node).update(ModelWithCoordinates, {
                atomicCoordinateFrame: {
                    elementCount: x.length,
                    time: { value: 0, unit: 'step' },
                    xyzOrdering: { isIdentity: true },
                    x: x,
                    y: y,
                    z: z
                }
            }).commit();
        }
    }

    // other methods to manupulate the structure. This could also be mapped directly to Python land

    constructor(public parent: PluginWrapper) {
    }
}
```

**Comment by hainm** [comment]
@dsehnal I prefer not to keep track of what to be added/removed and prefer molstar does that.
e.g: what if user creates the `view` in notebook and add the new structure in the molstar's GUI. Then the view has not knowledge of that. 

But I will look at your example carefully if I can reuse some of it. Thanks.
`selectQ(q => q.byRef(this.model.ref)` -> So basically I need to figure out how to get the `ref` in `this.model.ref` from a given `plugin` with a given structure (model) index.

**Comment by dsehnal** [comment]
Well, if the user modifies the state too much, you would not be able to recover from that anyway. The solution I suggested gives the user a handle to the molecule object to tweak it from the code as well. But up to you of course.

**Comment by hainm** [comment]
Thanks @dsehnal for the suggestion. I will keep them in mind. (I am trying to keep the same design pattern as we did for `NGL` in nglview (and I included `mostar` as another renderer for nglview).
